In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('../data/breadt_football_game_list.csv', parse_dates=['time'])
df.head()

id  matchid status game  turn home_team visit_team  gs  gd  gn       time  \
0   1  1836165      完  天皇杯    决赛      横滨水手       广岛三箭   2   0   2 2014-01-01   
1   2  1820486      完   澳超  第13轮       西悉尼        惠灵顿   1   3   4 2014-01-01   
2   3  1765894      完   英超  第20轮       斯旺西         曼城   2   3   5 2014-01-01   
3   4  1765886      完   英超  第20轮       阿森纳       加的夫城   2   0   2 2014-01-01   
4   5  1765887      完   英超  第20轮       水晶宫        诺维奇   1   1   2 2014-01-01   

   result  win_bet_return  draw_bet_return  lose_bet_return  
0       3            2.58             2.97             2.74  
1       0            1.41             4.15             7.16  
2       0            5.31             3.90             1.61  
3       3            1.25             5.86            11.58  
4       1            2.31             3.17             3.14

In [ ]:
df['game'].drop_duplicates()

In [34]:
high_partition = df[df['win_bet_return'] < 2]
high_partition.head()

id  matchid status game  turn home_team visit_team  gs  gd  gn       time  \
1    2  1820486      完   澳超  第13轮       西悉尼        惠灵顿   1   3   4 2014-01-01   
3    4  1765886      完   英超  第20轮       阿森纳       加的夫城   2   0   2 2014-01-01   
6    7  1765889      完   英超  第20轮       利物浦        赫尔城   2   0   2 2014-01-01   
11  12  1768225      完   英冠  第24轮       伯明翰       巴恩斯利   1   1   2 2014-01-01   
13  14  1768223      完   英冠  第24轮       布赖顿       伯恩茅斯   1   1   2 2014-01-01   

    result  win_bet_return  draw_bet_return  lose_bet_return  
1        0            1.41             4.15             7.16  
3        3            1.25             5.86            11.58  
6        3            1.29             5.46             9.85  
11       1            1.81             3.45             4.16  
13       1            1.81             3.44             4.23

In [9]:
len(high_partition[high_partition['result'] == 3]) / len(high_partition)

0.6168565712794642

In [10]:
high_partition_draw = df[df['draw_bet_return'] < 2]
high_partition_draw.head()

id  matchid status  game turn home_team visit_team  gs  gd  gn  \
23653  23654  2071948      完  解放者杯  分组赛       蒙国民       罗萨里奥   0   2   2   
32475  32476  2191138      完   非洲杯  分组赛        埃及         加纳   1   0   1   

            time  result  win_bet_return  draw_bet_return  lose_bet_return  
23653 2016-04-15       0            4.51             1.84             3.36  
32475 2017-01-26       3            3.15             1.86             4.36

In [11]:
high_partition_lose = df[df['lose_bet_return'] < 2]
high_partition_lose.head()

id  matchid status  game  turn home_team visit_team  gs  gd  gn  \
2    3  1765894      完    英超  第20轮       斯旺西         曼城   2   3   5   
8    9  1765892      完    英超  第20轮      斯托克城        埃弗顿   1   1   2   
18  19  1768218      完    英冠  第24轮      米尔沃尔       莱切斯特   1   3   4   
32  33  1770722      完    英甲  第24轮      特兰米尔         狼队   1   1   2   
40  41  1834441      完  球会友谊    一月      巴黎圣曼         皇马   0   1   1   

         time  result  win_bet_return  draw_bet_return  lose_bet_return  
2  2014-01-01       0            5.31             3.90             1.61  
8  2014-01-01       1            4.11             3.37             1.90  
18 2014-01-01       0            4.01             3.44             1.85  
32 2014-01-01       1            5.08             3.62             1.65  
40 2014-01-02       0            3.42             3.51             1.95

In [12]:
len(high_partition_lose[high_partition_lose['result'] == 0]) / len(high_partition_lose)

0.605312794718642

In [14]:
high_partition_lose['lose_bet_return'].mean()

1.6354746934926123

In [49]:
d = high_partition_lose[high_partition_lose['game'] == '英超']
d = d[d['time'] > '2019-01-01']
len(d[d['result'] == 0]) / len(d)

0.625

In [50]:
t = high_partition[high_partition['game'] == '英超']
t = t[t['time'] > '2019-01-01']
len(t[t['result'] == 3]) / len(t)

0.7941176470588235

In [51]:
t['win_bet_return'].mean()

1.4605882352941177

## 1.x : 4.x : 8.x

In [633]:
game_name = '澳超'

In [634]:
m = df[ (df['win_bet_return'] < df['lose_bet_return']) & (df['draw_bet_return'] < df['lose_bet_return']) & ((df['win_bet_return'] + df['draw_bet_return']) < df['lose_bet_return'])]
m  = m[m['win_bet_return'] < 2]
m = m[m['game'] == game_name]

bet_win_rate = 1-len(m[m['result'] == 0]) / len(m)
win_rate = len(m[m['result'] == 3]) / len(m)
draw_rate = len(m[m['result'] == 1]) / len(m)

bet_win_rate, win_rate, draw_rate

(0.858974358974359, 0.7564102564102564, 0.10256410256410256)

In [635]:
m['win_bet_return'].mean() * win_rate, m['draw_bet_return'].mean() * draw_rate, len(m)

(1.050440499671269, 0.48309007232084156, 78)

In [636]:
m = df[ (df['draw_bet_return'] < df['win_bet_return']) & (df['lose_bet_return'] < df['win_bet_return']) & ((df['draw_bet_return'] + df['lose_bet_return']) < df['win_bet_return'])]
m  = m[m['lose_bet_return'] < 2]
m = m[m['game'] == game_name]

bet_win_rate = 1-len(m[m['result'] == 3]) / len(m)
win_rate = len(m[m['result'] == 0]) / len(m)
draw_rate = len(m[m['result'] == 1]) / len(m)

bet_win_rate, win_rate, draw_rate

(1.0, 0.8571428571428571, 0.14285714285714285)

In [637]:
m['lose_bet_return'].mean() * win_rate, m['draw_bet_return'].mean() * draw_rate, len(m)

(1.2355102040816326, 0.6475510204081633, 7)

+ 澳超(1-1.0504|2-1.2355)
+ 美职(1-1.0054|2-1.5066)
+ 意甲(2-1.0117)
+ 俄超(2-1.1003)
+ 法甲(1-1.0147|2-1.0453)
+ 葡超(1-1.1021)
+ 法乙(2-1.5079)

## 高收益方案 No.1 Alpha

主队输德概率大于2，并且主队平的概率小于主队赢的概率 + 主队输的概率小于主队平的概率

lose_bet_return < draw_bet_return < win_bet_return

+ **意甲 - LOSE - 0.5826 - 1.2612 - 115**
+ **葡超 - DRAW - 0.3888 - 1.2322 - 54**
+ 澳超 - LOSE - 0.5000 - 1.0250 - 6
+ 澳超 - DRAW - 0.3333 - 1.1111 - 6
+ 美职 - DRAW - 0.5000 - 1.0474 - 2
+ 美职 - DRAW - 0.5000 - 1.6050 - 2
+ 西甲 - LOSE - 0.4875 - 1.0604 - 80
+ **德甲 - LOSE - 0.6250 - 1.3091 - 24**
+ **德乙 - LOSE - 0.5277 - 1.1279 - 36**
+ 荷甲 - LOSE - 0.8000 - 1.6384 - 5
+ 荷乙 - LOSE - 1.0000 - 2.7100 - 1
+ **英超 - WIN - 0.38356 - 1.3448 - 73**
+ **英甲 - LOSE - 0.5230 - 1.1046 - 65**
+ 挪超 - LOSE - 0.5000 - 1.0450 - 8
+ **韩K联 - LOSE - 0.6176 - 1.3119 - 34**
+ 瑞典超 - DRAW - 0.3200 - 1.0462 - 25
+ 挪超 - LOSE - 0.5000 - 1.0450 - 8

In [591]:
game_name = '韩K联'

In [592]:
m = df[ (df['draw_bet_return'] < df['win_bet_return']) & (df['lose_bet_return'] < df['draw_bet_return']) ]
m = m[m['lose_bet_return'] > 2]
m = m[m['game'] == game_name]

lose_rate = len(m[m['result'] == 0]) / len(m)
draw_rate = len(m[m['result'] == 1]) / len(m)
win_rate = len(m[m['result'] == 3]) / len(m)

lose_rate, draw_rate, win_rate

(0.6176470588235294, 0.2647058823529412, 0.11764705882352941)

In [593]:
m['lose_bet_return'].mean() * lose_rate, m['draw_bet_return'].mean() * draw_rate, m['win_bet_return'].mean() * win_rate, len(m)

(1.3119550173010381, 0.8321885813148789, 0.39262975778546716, 34)

In [528]:
m.head()

id  matchid status game  turn home_team visit_team  gs  gd  gn  \
503    504  1796086      完   意甲  第21轮      卡利亚里       AC米兰   1   2   3   
866    867  1796128      完   意甲  第24轮      卡塔尼亚        拉齐奥   3   1   4   
1299  1300  1796173      完   意甲  第26轮       萨索洛        帕尔马   0   1   1   
1319  1320  1796172      完   意甲  第26轮      AC米兰       尤文图斯   0   2   2   
1743  1744  1796236      完   意甲  第28轮       维罗纳       国际米兰   0   2   2   

           time  result  win_bet_return  draw_bet_return  lose_bet_return  
503  2014-01-26       0            3.79             3.33             2.05  
866  2014-02-16       3            3.37             3.15             2.18  
1299 2014-03-02       0            3.45             3.25             2.12  
1319 2014-03-03       0            3.60             3.28             2.06  
1743 2014-03-16       0            3.46             3.28             2.10

## 尝试 - 高收益方案

win_bet_return < lose_bet_return < draw_bet_return

+ 意甲 - DRAW - 0.3377 - 1.0980
+ 法甲 - WIN  - 0.4166 - 1.0417
+ 澳超 - LOSE - 0.3380 - 1.0107
+ 葡超 - WIN  - 0.4180 - 1.0236

In [668]:
game_name = '葡超'

In [669]:
m = df[ (df['win_bet_return'] < df['lose_bet_return']) & (df['lose_bet_return'] < df['draw_bet_return']) ]
m = m[m['lose_bet_return'] > 2]
m = m[m['game'] == game_name]

lose_rate = len(m[m['result'] == 0]) / len(m)
draw_rate = len(m[m['result'] == 1]) / len(m)
win_rate = len(m[m['result'] == 3]) / len(m)

lose_rate, draw_rate, win_rate

(0.2655367231638418, 0.3163841807909605, 0.4180790960451977)

In [670]:
m['lose_bet_return'].mean() * lose_rate, m['draw_bet_return'].mean() * draw_rate, m['win_bet_return'].mean() * win_rate, len(m)

(0.7705815697915669, 0.9834543075106131, 1.023632417249194, 177)

In [661]:
m.head()

id  matchid status game  turn home_team visit_team  gs  gd  gn  \
75    76  1783340      完   西甲  第18轮      巴利亚多        贝蒂斯   0   0   0   
82    83  1783339      完   西甲  第18轮      阿梅里亚       格拉纳达   3   0   3   
105  106  1783341      完   西甲  第18轮      奥萨苏纳       西班牙人   1   0   1   
106  107  1783338      完   西甲  第18轮      皇家社会       毕尔巴鄂   2   0   2   
224  225  1783351      完   西甲  第19轮       莱万特        马拉加   1   0   1   

          time  result  win_bet_return  draw_bet_return  lose_bet_return  
75  2014-01-05       1            2.36             3.25             2.95  
82  2014-01-05       3            2.43             3.19             2.89  
105 2014-01-06       3            2.38             3.14             3.01  
106 2014-01-06       3            2.16             3.32             3.24  
224 2014-01-13       3            2.56             3.17             2.73

## 高收益方案 No.2 beta

draw_bet_return > win_bet_return > lose_bet_return

+ **德甲 - WIN - 0.3799 - 1.1350 - 229**
+ 德乙 - DRAW - 0.3166 - 1.0246 - 240
+ **西甲 - DRAW - 0.3473 - 1.1421 - 190**
+ 法甲 - LOSE - 0.4022 - 1.0017 - 174
+ 日职 - LOSE - 0.4542 - 1.066 - 306
+ 英超 - DRAW - 0.3208 - 1.0590 - 187
+ 英甲 - WIN - 0.7500 - 2.1037 - 4
+ **韩K联 - DRAW - 0.3636 - 1.1632 - 143**
+ 澳超 - LOSE - 0.4409 - 1.0220 - 127
+ **美职 - WIN - 0.4503 - 1.2673 - 131**
+ 葡超 - WIN - 0.3709 - 1.0672 - 124
+ 挪超 - DRAW - 0.3053 - 1.0235 - 167

In [585]:
game_name = '德甲'

In [586]:
m = df[ (df['draw_bet_return'] > df['win_bet_return']) & (df['lose_bet_return'] < df['win_bet_return']) ]
m = m[m['lose_bet_return'] > 2]
m = m[m['game'] == game_name]

lose_rate = len(m[m['result'] == 0]) / len(m)
draw_rate = len(m[m['result'] == 1]) / len(m)
win_rate = len(m[m['result'] == 3]) / len(m)

win_rate, draw_rate, lose_rate

(0.3799126637554585, 0.27074235807860264, 0.34934497816593885)

In [587]:
m['win_bet_return'].mean() * win_rate, m['draw_bet_return'].mean() * draw_rate, m['lose_bet_return'].mean() * lose_rate, len(m)

(1.1350264106329018, 0.9202993840697166, 0.8237524074674395, 229)

In [588]:
m.head()

id  matchid status game  turn home_team visit_team  gs  gd  gn  \
515    516  1766338      完   德甲  第18轮        汉堡      沙尔克04   0   3   3   
610    611  1766361      完   德甲  第20轮       弗赖堡       霍芬海姆   1   1   2   
814    815  1766365      完   德甲  第21轮      不伦瑞克         汉堡   4   2   6   
993    994  1766374      完   德甲  第22轮       弗赖堡       奥格斯堡   2   4   6   
1211  1212  1766387      完   德甲  第23轮       不来梅         汉堡   1   0   1   

           time  result  win_bet_return  draw_bet_return  lose_bet_return  
515  2014-01-27       0            2.67             3.44             2.48  
610  2014-02-08       1            2.81             3.38             2.39  
814  2014-02-15       3            2.69             3.26             2.54  
993  2014-02-22       0            3.21             3.25             2.21  
1211 2014-03-01       3            2.64             3.37             2.52

## 尝试 - 高收益方案 No.3


In [605]:
game_name = '英甲'

In [606]:
m = df[ (df['lose_bet_return'] > df['draw_bet_return']) & (df['draw_bet_return'] > df['win_bet_return']) ]
m = m[m['win_bet_return'] > 2]
m = m[m['game'] == game_name]

lose_rate = len(m[m['result'] == 0]) / len(m)
draw_rate = len(m[m['result'] == 1]) / len(m)
win_rate = len(m[m['result'] == 3]) / len(m)

win_rate, draw_rate, lose_rate

(0.3671875, 0.3515625, 0.28125)

In [607]:
m['win_bet_return'].mean() * win_rate, m['draw_bet_return'].mean() * draw_rate, m['lose_bet_return'].mean() * lose_rate, len(m)

(0.7708212280273437, 1.1460525512695312, 0.9646655273437499, 256)